[lldb/formatter] Add Swift.UnsafeRawBufferPointer data formatter

medismailben opened the pull request:
This patch adds support for Swift's UnsafeRawBufferPointer and
UnsafeMutableRawBufferPointer data formatting.

This commit also introduces a new SwiftUnsafeType factory to adapt and
use both Unsafe{,Raw}BufferPointer with the pre-existing data formatter.

rdar://63946760

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

medismailben commented:
@swift-ci test

jimingham commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -74,9 +77,20 @@ bool SwiftUnsafeBufferPointer::Update() {
+}
+
+lldb::addr_t SwiftUnsafeBufferPointer::GetAddress(llvm::StringRef child_name) {
+  ConstString g_child_name(child_name);
```
g_something usually means the thing is a static, either local to the function or a file static.  This is neither, so g_ isn't an appropriate name.

jimingham commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -89,35 +103,104 @@ bool SwiftUnsafeBufferPointer::Update() {
-    return false;
-
-  m_elem_type = argument_type;
+  if (argument_type.IsValid())
```
Why is it okay to keep going here if you can't get a valid type for the contents of the UnsafeBufferPointer?  If there is a good reason, it would be good to explain it here.

medismailben replied:
`UnsafeRawBufferPointer` doesn't have a generic argument since by design, it's a raw byte view of the pointed object. Since `GetAddress` is used for both type, when it's called from the `UnsafeRawBufferPointer` helper class, the generic argument type will be invalid, which leaves `m_elem_type` uninitialized (invalid). Later on, in `UnsafeRawBufferPointer::Update`, I check if `m_elem_type` is invalid, in which case, I reconstruct the `Swift.UInt8` compiler type.

If you look at `UnsafeRawBufferPointer`'s layout, we can't get the generic argument type with the top-level ValueObject ("_position"). To get it, lldb need to look one level below, at the "some" ValueObject node.
That's the reason why I do it in `GetAddress`, on the way of fetching the "_pointerValue" address.

I should probably add a comment to explain this.

medismailben commented:
@swift-ci test Linux Platform

jimingham commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -149,7 +232,7 @@ class UnsafeBufferPointerSyntheticFrontEnd : public SyntheticChildrenFrontEnd {
   lldb::ByteOrder m_order;
 
-  SwiftUnsafeBufferPointer m_unsafe_ptr;
+  SwiftUnsafeBufferPointer *m_unsafe_ptr;
```
You are leaking this SwiftUnsafeBufferPointer now.  Probably better to make this a unique_ptr<SwiftUnsafeBufferPointer>, then it will get cleaned up automatically.

jimingham commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -89,35 +103,104 @@ bool SwiftUnsafeBufferPointer::Update() {
 
-  size_t count = swift_ubp.GetCount();
-  addr_t addr = swift_ubp.GetStartAddress();
+  if (!type_name.contains("Raw")) {
```
Be nice to unify this.  If you had a factory that returned the right kind of object as a SwiftUnsafeBufferPointerUP then you wouldn't need the if/else

jimingham commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -170,13 +252,18 @@ lldb_private::formatters::swift::UnsafeBufferPointerSyntheticFrontEnd::
   m_order = process_sp->GetByteOrder();
 
+  llvm::StringRef valobj_type_name(valobj_sp->GetTypeName().GetCString());
+  bool is_raw = valobj_type_name.contains("Raw");
```
It would be better to have a factory method that makes the right version as a UP.  Then you could set it into your ivar here and in the Summary you could also use that to avoid the ugly if (Raw) {make_raw; do stuff; } else {make_not_raw; do same stuff } construct.

medismailben replied:
Will do!

jimingham commented on lldb/test/API/functionalities/data-formatter/swift-unsafe/main.swift:
```diff
@@ -117,7 +117,43 @@ func main() {
+    //%            '\[1\] = 0',
+    //%            ])
+  }
+
```
Be good to test that "frame var mutrawbuf[0]" also works.

medismailben replied:
I'll add a check for that

jimingham reviewed (commented):
This looks pretty good.  The synthetic provider is leaking its SwiftUnsafeBufferPointer object, and I think you can isolate the choice to make the raw vrs. not raw difference.  Once you've made the right object only call base class methods so you should always handle the base class and let a factory make the right one for you.

medismailben commented:
Thanks for the feedbacks @jimingham ! Indeed the factory approach would be a good way to manage both types. I'll wait to see if @dcci and @fredriss could think of other improvements I could make then update the PR.

SNiTEBoBy commented:
Merge pull requested

medismailben commented:
I updated the patch addressing @jimingham comments:

- Added a `SwiftUnsafeType` factory for UnsafeBufferPointer and UnsafeRawBufferPointer
- Used `std::unique_ptr` instead of a raw pointer.
- Added the test Jim requested.

medismailben commented:
@swift-ci test

medismailben commented:
@swift-ci test

jimingham commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -74,50 +114,108 @@ bool SwiftUnsafeBufferPointer::Update() {
 
+std::unique_ptr<SwiftUnsafeType> SwiftUnsafeType::Create(ValueObject &valobj) {
+  llvm::StringRef valobj_type_name(valobj.GetTypeName().GetCString());
+  bool is_raw = valobj_type_name.contains("Raw");
```
The formatter matching looks through typedefs.  So your valobj might have matched the formatter, but the top-level typename of this value object might be a typedef (and thus not contain "Raw").  The matching needs to also look through typedefs to the resolved type or you might get the wrong formatter.

jimingham reviewed (commented):
That looks better to me.  

You have to handle the case where the valobj is a typedef to UnsafeBufferPointer or UnsafeRawBufferPointer.  Just strip off typedef's before looking for the "Raw" in the typename.

You might also want to think a bit about "if this were to fail in the field, where would I wish I had put in logging to help me figure out what went wrong" and if there are any obvious places add some "formatter" channel logging

dcci commented on lldb/test/API/functionalities/data-formatter/swift-unsafe/main.swift:
```diff
@@ -28,9 +28,9 @@ struct Plane: Flyable {
 
 class Number<T:Numeric> {
-  
+
```
spurious.

dcci commented on lldb/test/API/functionalities/data-formatter/swift-unsafe/main.swift:
```diff
@@ -50,7 +50,7 @@ func main() {
     //%            ])
   }
-  
+
```
etc...

dcci commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -74,50 +114,108 @@ bool SwiftUnsafeBufferPointer::Update() {
 
-  if (!swift_ubp.Update())
-    return false;
+  std::unique_ptr<SwiftUnsafeType> unsafe_type =
```
clang-format?

dcci commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -74,50 +114,108 @@ bool SwiftUnsafeBufferPointer::Update() {
+
+  if (!m_elem_type.IsValid()) {
+    auto ast_context = m_valobj.GetScratchSwiftASTContext();
+    if (!ast_context)
```
you need to grab the ASTContext lock & make sure the ASTContext has no errors.

dcci commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -9,24 +9,64 @@
+  SwiftUnsafeType(ValueObject &valobj);
+  addr_t GetAddress(llvm::StringRef child_name);
+
+protected:
```
protected twice?

dcci reviewed (changes requested):
A round of comments.

medismailben commented:
@swift-ci test

jimingham commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -74,50 +118,120 @@ bool SwiftUnsafeBufferPointer::Update() {
+  //     ▿ some : UnsafeRawPointer<Int>
+  //       - pointerValue : Int
+  //
+  // The structure has 2 Optional UnsafePointer to the buffer's start address
```
UnsafePointer -> UnsafePointers

jimingham commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -74,50 +118,120 @@ bool SwiftUnsafeBufferPointer::Update() {
+  //
+  // The structure has 2 Optional UnsafePointer to the buffer's start address
+  // and end address. To access the pointer address, lldb unfolds every value
+  // object child until reaching `pointerValue`.
```
value object -> ValueObject?

jimingham commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -74,50 +118,120 @@ bool SwiftUnsafeBufferPointer::Update() {
+
+  opaque_compiler_type_t qual_type = type.GetOpaqueQualType();
+
+  if (type_system->IsTypedefType(qual_type))
```
GetTypedefType only undoes one level of typedef'ing.  So the result might still be a typedef.  You probably need to loop over IsTypedefType till you get to the non-typedef'ed version.

jimingham commented on lldb/source/Target/SwiftLanguageRuntimeNames.cpp:
```diff
@@ -393,13 +393,15 @@ SwiftLanguageRuntime::DemangleSymbolAsString(StringRef symbol, DemangleMode mode
     options.DisplayModuleNames = true;
     options.ShowPrivateDiscriminators = false;
     options.DisplayExtensionContexts = false;
+    options.DisplayLocalNameContexts = false;
```
Why was this necessary?

medismailben replied:
Previously, when a type alias was created inside a closure, the Swift demangler included the closure signature in the demangled type:

```
ByteBuffer #1 in closure #6 (Swift.UnsafeBufferPointer<Swift.UInt8>) -> () in main.main() -> ()
```

In my case, this was causing an assertion when fetching the Typedef CompilerType, because both signature didn't match . This was fixed with #1331 and https://github.com/apple/swift/pull/32336.

Now, by setting `DemangleOptions::DisplayLocalNameContexts` to `false`, the demangler only returns the actual type (`ByteBuffer`), avoiding the assertion.

jimingham reviewed (commented):
A few nits and one annoyance w.r.t. typedef's.

medismailben commented:
Addressed final round of comments:
- Handled typedef chaining and added test for it
- Fixed typos.
- Added logging.

medismailben committed:
[lldb/formatter] Add Swift.UnsafeRawBufferPointer data formatter

This patch adds support for Swift's UnsafeRawBufferPointer and
UnsafeMutableRawBufferPointer data formatting.

This commit also introduces a new SwiftUnsafeType factory to adapt and
use both Unsafe{,Raw}BufferPointer with the pre-existing data formatter.

rdar://63946760

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

medismailben commented:
@swift-ci test

medismailben commented:
@swift-ci test
